docs/design/validation.md

```diff
@@ -0,0 +1,1989 @@
+
+```terraform
+$ cat main.tf
+resource "random_pet" "example" {
+  length = 2
+}
+
+$ terraform plan
+
+Terraform used the selected providers to generate the following execution plan. Resource actions are
+indicated with the following symbols:
+  + create
+
+Terraform will perform the following actions:
+
+  # random_pet.example will be created
+  + resource "random_pet" "example" {
+      + id        = (known after apply)
+      + length    = 2
+      + separator = "-"
+    }
+
+Plan: 1 to add, 0 to change, 0 to destroy.
+```
+
+This plan is surfaced to providers in a machine readable manner through the [Terraform Plugin Protocol](#terraform-plugin-protocol) which is described later.
+
+### Terraform Providers
+
+Terraform Providers are a form of Terraform plugins, which are gRPC (formerly `net/rpc`) server processes that are lifecycle managed by Terraform CLI. Providers implement [managed resources](https://www.terraform.io/docs/language/resources/) and [data sources](https://www.terraform.io/docs/language/data-sources/). Often a managed resource is just called a "resource" although in the underlying implementation details a "resource" may refer to either for legacy reasons as seen later in the previous provider framework.
+
+From a configuration standpoint, Terraform implements the concepts of `provider`, `resource`, and `data` (source) configurations, while providers implement the details inside those configurations in what is called "schema" information. The schema defines attribute naming, types, and behaviors for consumption by the framework and Terraform CLI.
+
+To visualize the difference in a configuration:
+
+```terraform
+# Provider configuration
+provider "example" { # Defined by the configuration language
+  # Defined by the provider schema
+  input = "" # a required or optional string attribute named input
+}
+
+# Resource configuration
+resource "example_thing" "example" { # Defined by the configuration language
+  # Defined by the resource schema
+  input = 123 # a required or optional number attribute named input
+}
+
+# Data source configuration
+data "example_thing" "example" { # Defined by the configuration language
+  # Defined by the data source schema
+  input = true # a required or optional boolean attribute named input
+}
+```
+
+Terraform supports the following validation for provider implementations:
+
+- Provider configurations
+- Resource configurations (and configurations during plans)
+- Data Source configurations (and configurations during plans)
+
+Within these, there are two types of validation:
+
+- Single attribute value validation (e.g. string length)
+- Multiple attribute validation (e.g. attributes or attribute values that conflict with each other)
+
+There is no difference between these types of validation to Terraform, as Terraform just works with errors and warnings being returned from providers, but the previous framework surfaced these different conceptually.
```

Comment:
> these different conceptually

Maybe "these concepts differently"?

Reply:
Yes, indeed. Thank you for this suggestion.